Removal of the layer count restriction

gchoqueux opened the pull request:
- See proposal  (#2614 )  

**WARNING : There’s still unresolved bugs:**
* tiles disappear because of an improperly initialized bbox.
* Opacity, vibility and effects haven’t been reconnected yet.


## Description

Replace all textures array by just 2 textures, **map and displacement map**. All array textures are flatten in these maps.  

## Motivation and Context

* removal of the layer count restriction
* code simplification
* first step : 
    * to replace `LayeredMaterial`  with any Three.js materials
    * to support any raster re-projection

Neptilo commented on packages/Main/src/Renderer/LayeredMaterial.ts:
```diff
@@ -94,86 +94,35 @@ const rtCache = new LRUCache<string, THREE.WebGLArrayRenderTarget>({
+
+        drawMap(renderTarget, tiles, renderer, extent);
+
+        rtCache.set(textureSetId, renderTarget);
```
I was caching array render targets instead of array textures because it was impossible to extract the textures from the RTs and manage their memory separately, but maybe now that you're using simple render targets, this becomes possible and we can just maintain a cache of textures directly, and only use one RT for everything?

Neptilo commented on packages/Main/src/Renderer/WebGLComposer.ts:
```diff
@@ -1,106 +1,46 @@
     renderer.setRenderTarget(previousRenderTarget);
+    renderer.setClearAlpha(a);
 
     return renderTarget;
```
The returned value is not used anymore and the documented type is now incorrect

gchoqueux commented on packages/Main/src/Renderer/LayeredMaterial.ts:
```diff
@@ -94,86 +94,35 @@ const rtCache = new LRUCache<string, THREE.WebGLArrayRenderTarget>({
+
+        rtCache.set(textureSetId, renderTarget);
+    } else {
+        drawMap(renderTarget, tiles, renderer, extent);
```
Yes, it's temporary. I'll add a mechanism to update the texture and parent texture when the opacity changes.

Neptilo commented on packages/Main/src/Renderer/ProjectionMaterials.js:
```diff
@@ -0,0 +1,110 @@
+    uniforms: {
+        // This uniform will be updated with each source 2D texture
+        map: { value: null },
+        texture_footprint: { value: new THREE.Vector2() },
```
Could you comment what these properties represent?

gchoqueux commented on packages/Main/src/Renderer/ProjectionMaterials.js:
```diff
@@ -0,0 +1,110 @@
+        uniform vec4 offsetScale;
+        uniform float opacity;
+
+        const float CTOYL = 20037508.34 / 3.141592653589793 * 0.69314718;
```
i'll add some comment
* 20037508.34 : max projection in speudo mercator
* 0,69... : Ln(2)

Neptilo replied:
What do you mean by "max projection"?

Neptilo commented on packages/Main/src/Converter/convertToTile.ts:
```diff
@@ -89,8 +89,27 @@ export default {
                 tile.material.setUniform('geoidHeight', geoidHeight);
+
+                const parent_uniforms = parent.material.uniforms;
+                const child_uniforms = tile.material.uniforms;
```
Use camel-case to keep coding style consistent.

Neptilo commented on packages/Main/src/Converter/convertToTile.ts:
```diff
@@ -89,8 +89,27 @@ export default {
+                const parent_uniforms = parent.material.uniforms;
+                const child_uniforms = tile.material.uniforms;
+
+                if (parent.material.uniforms.map.value) {
```
You can use the previously defined `parent_uniforms` here.

Neptilo commented on packages/Main/src/Converter/convertToTile.ts:
```diff
@@ -89,8 +89,27 @@ export default {
+                if (parent.material.uniforms.map.value) {
+                    const extentParent = parent_uniforms.map.value.extent;
+                    child_uniforms.map.value = parent_uniforms.map.value;
+                    child_uniforms.colorOffsetScales.value = tile.extent.offsetToParent(extentParent);
```
Minor: fix warnings for line lengths over 100 characters

Neptilo commented on packages/Main/src/Converter/convertToTile.ts:
```diff
@@ -89,8 +89,27 @@ export default {
+                    tile.setBBoxZ({ min: rasterElevationNode.min, max: rasterElevationNode.max, scale: rasterElevationNode.layer.scale, geoidHeight });
+                } else {
+                    tile.setBBoxZ({ min: parent.obb.z.min, max: parent.obb.z.max, geoidHeight });
+                }
```
An explanatory comment for all this added code would be nice.

Neptilo commented on packages/Main/src/Renderer/Shader/TileFS.glsl:
```diff
@@ -31,30 +31,17 @@ void main() {
+        gl_FragColor.rgb = mix(gl_FragColor.rgb, color.rgb, color.a);
     }
-  #endif
+    #endif
```
Minor: make sure to keep indentation consistent

Neptilo commented on packages/Main/src/Renderer/LayeredMaterial.ts:
```diff
@@ -90,88 +90,38 @@ const defaultStructLayers: Readonly<{
     }
 
-    const cachedRT = renderTargetCache?.get(textureSetId);
+    const textureSetId = extent.toString();
```
If the cache ids are only based on the extents, will we have a problem if there are multiple views? Different textures can be used for the same extent but in different views, e.g. a globe + a minimap.

Neptilo commented on packages/Main/src/Renderer/LayeredMaterial.ts:
```diff
@@ -482,45 +422,28 @@ export class LayeredMaterial extends THREE.ShaderMaterial {
         return {
             layers: this.uniforms[`${type}Layers`],
-            textures: this.uniforms[`${type}Textures`],
+            textures: type == 'color' ? this.uniforms.map : this.uniforms.dislacementMap,
```
"dislacementMap": typo.
Could we modify the type of `uniforms` to prevent this kind of mistake? (I'm not sure myself - may be overkill)

Neptilo commented on packages/Main/src/Renderer/LayeredMaterial.ts:
```diff
@@ -482,45 +422,28 @@ export class LayeredMaterial extends THREE.ShaderMaterial {
-     * @param renderer - The renderer used to render textures into arrays.
-     */
-    public updateLayersUniforms(renderer: THREE.WebGLRenderer): void {
+    public updateLayersUniforms(renderer: THREE.WebGLRenderer, extent: Extent): void {
```
It's always good to keep function documentation.

Neptilo commented on packages/Main/src/Renderer/ProjectionMaterials.js:
```diff
@@ -0,0 +1,110 @@
+
+        float toY(float lat) {
+            float s = sin(lat);
+            return 0.5 * log2((1.0 + s) / (1.0 - s)) * CTOYL;
```
Where does this formula come from?

Neptilo commented on packages/Main/src/Renderer/ProjectionMaterials.js:
You could use this opportunity to already write this new file in TypeScript.

Neptilo commented on packages/Main/src/Renderer/WebGLComposer.ts:
```diff
@@ -1,117 +1,52 @@
+
+const quad = new THREE.Mesh(geometry, materialUnit);
 
+export function drawMap(
```
Can you document the function?

Neptilo commented on packages/Main/src/Renderer/WebGLComposer.ts:
```diff
@@ -1,117 +1,52 @@
-        depthBuffer: false, // No depth buffer needed for simple 2D texture copy
-    });
-    const arrayTexture = renderTarget.texture;
+const forwarding = (from, to) => ((from == to) ? 'transformUnit' : `${from} => ${to}`);
```
Can you document what `materials` contains and give `forwarding` a more explicit name? For example, "makeProjectionId" if you think that makes sense.

Neptilo commented on packages/Main/src/Renderer/ProjectionMaterials.js:
Could you document what the purpose of these materials is? As I understand it: running projection computations on the GPU?

gchoqueux committed:
fix(test): issues from refactoring and rebasing

gchoqueux committed:
fix(test): issues from refactoring and rebasing

gchoqueux committed:
example(tms): add example 2d and 3d

gchoqueux committed:
refacto(LayeredMaterial): one color map and one displacementMap.

gchoqueux committed:
refacto(LayeredMaterial): one color map and one displacementMap.

gchoqueux committed:
refacto(composer): temp

gchoqueux committed:
refacto(composer): temp

gchoqueux committed:
refacto(composer): add pass

gchoqueux committed:
refacto(composer): add pass

gchoqueux committed:
refacto(RasterTile): simplify raster layer loading

BREACKING CHANGES: remove pdateLayeredMaterialNodeImagery, updateLayeredMaterialNodeElevation

gchoqueux committed:
refacto(examples): use await top level to fetch

gchoqueux committed:
refacto(RasterTile): simplify raster layer loading

BREACKING CHANGES: remove pdateLayeredMaterialNodeImagery, updateLayeredMaterialNodeElevation

gchoqueux committed:
fix(TiledGeometryLayer): [WIP][PROPOSAL] avoid to subdivide node mesh if raster tiles aren't loaded

gchoqueux committed:
refacto(RasterLayer): update tile material uniform when textures are loaded
